### ISSUE TITLE
question: (Traefik + haproxy) - How to configure for DMS?

### ISSUE THREAD

**Issue by sanjuant** [docker-mailserver/docker-mailserver#3547]
### Subject

I would like some feedback concerning a use case

### Description

📝 Preliminary Checks

I tried searching for an existing issue and followed the [debugging docs](https://docker-mailserver.github.io/docker-mailserver/latest/config/debugging/) advice, but still need assistance.

👀 What Happened?

Hello, I've managed to set up the service locally on my server. But when I want to configure the haproxy to be able to use DMS behind a proxy, I can't configure it. I've followed the various issues that talk about this problem but I haven't managed to find an answer. 

When the server starts, I get the error `postconf: warning: unmatched service_name/type: "submissions/inet/smtpd_upstream_proxy_protocol=haproxy"`
```
[   INF   ]  Welcome to docker-mailserver 12.1.0
[   INF   ]  Checking configuration
[   INF   ]  Configuring mail server
[ WARNING ]  (Rspamd setup) Running Amavis/SA & Rspamd at the same time is discouraged
[ WARNING ]  (Rspamd setup) Running OpenDMARC & Rspamd at the same time is discouraged - we recommend Rspamd for DMARC checks (enabled with Rspamd by default)
[ WARNING ]  (Rspamd setup) Running policyd-spf & Rspamd at the same time is discouraged - we recommend Rspamd for SPF checks (enabled with Rspamd by default)
postconf: warning: unmatched service_name/type: "submissions/inet/smtpd_upstream_proxy_protocol=haproxy"
[   INF   ]  Starting daemons
[   INF   ]  mail.example.pro is up and running
Sep 25 21:58:59 mail amavis[681]: starting. /usr/sbin/amavisd-new at mail.example.pro amavisd-new-2.11.1 (20181009), Unicode aware, LC_CTYPE="C.UTF-8"
Sep 25 21:58:59 mail amavis[681]: perl=5.032001, user=, EUID: 109 (109);  group=, EGID: 111 111 (111 111)
Sep 25 21:58:59 mail amavis[681]: Net::Server: Group Not Defined.  Defaulting to EGID '111 111'
Sep 25 21:58:59 mail amavis[681]: Net::Server: User Not Defined.  Defaulting to EUID '109'
```

In the **haproxy_trusted_networks** variable in the dovecot conf I thought it was sufficient to specify the ip of the traefik network, but I'm not sure. Here  `172.21.0.1` ? I tried several ip like 172.23.0.1, 172.21.0.0/16, 172.23.0.0/16
![image](https://github.com/docker-mailserver/docker-mailserver/assets/27998877/41ba4f85-302f-448d-b53c-f953b1375988)

I tried a lot of different things before coming here to ask my question. I'm not very experienced in networking. Thank a lot for your help.

👟 Reproduction Steps
Use DMS behind Traefik reverse proxy

🐋 DMS Version
v12.1.0

💻 Operating System and Architecture
Linux sbserver 5.15.0-83-generic https://github.com/docker-mailserver/docker-mailserver/pull/92-Ubuntu SMP Mon Aug 14 09:30:42 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

⚙️ Container configuration files
traefik compose
```yaml
services:
  dockerproxy: # see https://github.com/Tecnativa/docker-socket-proxy
    container_name: dockerproxy
    image: ghcr.io/tecnativa/docker-socket-proxy:0.1.1 # this image is rather old but used to have a pinned version
                                                       # newer version is ghcr.io/tecnativa/docker-socket-proxy:edge
    restart: unless-stopped
    mem_limit: 2G
    cpus: 0.75
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock:ro   # never expose this container to the internet!
    environment:
      CONTAINERS: 1
    networks:
      - docker-proxynet # use only internal network

  traefik:
    container_name: traefik
    image: traefik:2.8 # use current version 2.8.x
    restart: unless-stopped
    read_only: true
    mem_limit: 2G
    cpus: 0.75
    depends_on:
      - dockerproxy
    security_opt:
      - no-new-privileges:true
    volumes:
      - /data/traefik/config/traefik.yaml:/etc/traefik/traefik.yaml:ro    # Static Traefik Configuration
      - /data/traefik/config/dynamic/:/etc/traefik/dynamic/:ro            # Folder to store dynamic configuration file provider
      - /data/traefik/config/authusers.txt:/etc/traefik/authusers.txt:ro  # userfile for basic auth
      - /data/traefik/config/acme/acme.json:/etc/traefik/acme.json        # certificate storage
    user: "2000:2000" # user traefik has to be created on the host system
                      # sudo useradd -u 2000 -M -s /usr/sbin/nologin traefik
    labels:
      - "traefik.enable=true"
      # expose traefik dashboard with TLS and basic auth
      # 1. create router
      - "traefik.http.routers.traefik.rule=Host(`traefik.example.pro`)"
      - "traefik.http.routers.traefik.service=api@internal"
      - "traefik.http.routers.traefik.tls.certresolver=tlschallenge"
      - "traefik.http.routers.traefik.entrypoints=web-secure"
      - "traefik.http.routers.traefik.middlewares=traefik-auth, secHeaders@file"
      # 2. create basic auth middleware
      - "traefik.http.middlewares.traefik-auth.basicauth.removeheader=true"
      - "traefik.http.middlewares.traefik-auth.basicauth.usersfile=/etc/traefik/authusers.txt"
      # this would be another possibility to create an catch-all on port 80 with redirection to https
      # but we connected the unencrypted entrypoint to the redirection in the static configuration:
      - "traefik.http.routers.http-catchall.rule=hostregexp(`{host:.+}`)"
      - "traefik.http.routers.http-catchall.entrypoints=web"
      - "traefik.http.routers.http-catchall.middlewares=redirect-to-https"
      - "traefik.http.middlewares.redirect-to-https.redirectscheme.scheme=https"
    networks:
      - traefik-servicenet
      - docker-proxynet
    ports:
      - "80:10080"  # use high ports inside the container so
      - "443:10443" # we don't need to be root to bind the ports
      - "25:25"
      - "465:465" 
      - "993:993"

networks:
  traefik-servicenet:
    external: true           # this network has to be created once before starting:
    name: traefik-servicenet # docker network create traefik-servicenet
  docker-proxynet:
    driver: bridge
    internal: true
```

dms compose
```yaml
services:
  mailserver:
    image: ghcr.io/docker-mailserver/docker-mailserver:latest
    container_name: mailserver
    # Provide the FQDN of your mail server here (Your DNS MX record should point to this value)
    hostname: mail.example.pro
    #domainname: 

    # ports:
    #   - "25:25"
    #   - "465:465"
    #   - "587:587"
    #   - "993:993"
    volumes:
      - /data/dms/mail-data/:/var/mail/
      - /data/dms/mail-state/:/var/mail-state/
      - /data/dms/mail-logs/:/var/log/mail/
      - /data/dms/config/:/tmp/docker-mailserver/
      - /etc/localtime:/etc/localtime:ro
      - /data/traefik/config/acme/acme.json:/etc/letsencrypt/acme.json:ro
    environment:
      - ENABLE_OPENDKIM=0
      - ENABLE_RSPAMD=1
      - ENABLE_CLAMAV=1
      - ENABLE_FAIL2BAN=1
      - SSL_TYPE=letsencrypt
      - SSL_DOMAIN=mail.example.pro
    cap_add:
      - NET_ADMIN # For Fail2Ban to work
    labels:
      - "traefik.enable=true"
      - "traefik.tcp.routers.smtp.rule=HostSNI(`*`)"
      - "traefik.tcp.routers.smtp.entrypoints=smtp"
      - "traefik.tcp.routers.smtp.service=smtp"
      - "traefik.tcp.services.smtp.loadbalancer.server.port=25"
      - "traefik.tcp.services.smtp.loadbalancer.proxyProtocol.version=1"
      - "traefik.tcp.routers.smtp-ssl.rule=HostSNI(`*`)"
      - "traefik.tcp.routers.smtp-ssl.tls=false"
      - "traefik.tcp.routers.smtp-ssl.entrypoints=smtp-ssl"
      - "traefik.tcp.routers.smtp-ssl.service=smtp-ssl"
      - "traefik.tcp.services.smtp-ssl.loadbalancer.server.port=465"
      - "traefik.tcp.services.smtp-ssl.loadbalancer.proxyProtocol.version=1"
      - "traefik.tcp.routers.imap-ssl.rule=HostSNI(`*`)"
      - "traefik.tcp.routers.imap-ssl.entrypoints=imap-ssl"
      - "traefik.tcp.routers.imap-ssl.service=imap-ssl"
      - "traefik.tcp.services.imap-ssl.loadbalancer.server.port=10993"
      - "traefik.tcp.services.imap-ssl.loadbalancer.proxyProtocol.version=2"
    restart: always
    networks:
      - traefik-servicenet

networks:
  traefik-servicenet:
    external: true
```

dovecot-quotas.cf
```
haproxy_trusted_networks = 172.21.0.1 # traefik-servicenet
haproxy_timeout = 3 sec
service imap-login {
  inet_listener imaps {
    haproxy = yes
    ssl = yes
    port = 10993
  }
}
```

postfix-master
```
submission/inet/smtpd_upstream_proxy_protocol=haproxy
submissions/inet/smtpd_upstream_proxy_protocol=haproxy
```

postfix-main
```
postscreen_upstream_proxy_protocol = haproxy
```

traefik.yaml
```yaml
# accessLog: {}
log:
  level: WARN
providers:
  docker:
    exposedByDefault: false
    endpoint: 'tcp://dockerproxy:2375'
    network: traefik-servicenet
  file:
    directory: /etc/traefik/dynamic
    watch: true
api:
  dashboard: true # if you don't need the dashboard disable it
entryPoints:
  web:
    address: ':10080' # will be routed to port 80, see docker-compose.yaml
    http:
      redirections:   # redirect entire entrypoint to https
        entryPoint:
          to: ':443'
          scheme: https
  web-secure:
    address: ':10443' # will be routed to port 443, see docker-compose.yaml
    http3:
      advertisedPort: 443
  smtp:
    address: ':25'
  smtp-ssl:
    address: ':465'
  imap-ssl:
    address: ':993'

certificatesResolvers:
  tlschallenge:
    acme:
      # uncomment next line for testing purpose
      # caServer: https://acme-staging-v02.api.letsencrypt.org/directory
      email: admin@example.pro
      storage: /etc/traefik/acme.json
      tlsChallenge: {}
experimental:
  http3: true
global:
  sendAnonymousUsage: false # set to false, if you don't want to send anonymous usage data
```

📜 Relevant log output
```
[   INF   ]  Welcome to docker-mailserver 12.1.0
[   INF   ]  Checking configuration
[   INF   ]  Configuring mail server
[ WARNING ]  (Rspamd setup) Running Amavis/SA & Rspamd at the same time is discouraged
[ WARNING ]  (Rspamd setup) Running OpenDMARC & Rspamd at the same time is discouraged - we recommend Rspamd for DMARC checks (enabled with Rspamd by default)
[ WARNING ]  (Rspamd setup) Running policyd-spf & Rspamd at the same time is discouraged - we recommend Rspamd for SPF checks (enabled with Rspamd by default)

postconf: warning: unmatched service_name/type: "submissions/inet/smtpd_upstream_proxy_protocol=haproxy"

[   INF   ]  Starting daemons
[   INF   ]  mail.example.pro is up and running
Sep 25 23:11:27 mail amavis[684]: starting. /usr/sbin/amavisd-new at mail.example.pro amavisd-new-2.11.1 (20181009), Unicode aware, LC_CTYPE="C.UTF-8"
Sep 25 23:11:27 mail amavis[684]: perl=5.032001, user=, EUID: 109 (109);  group=, EGID: 111 111 (111 111)
Sep 25 23:11:27 mail amavis[684]: Net::Server: Group Not Defined.  Defaulting to EGID '111 111'
Sep 25 23:11:27 mail amavis[684]: Net::Server: User Not Defined.  Defaulting to EUID '109'
Sep 25 23:11:27 mail amavis[684]: No ext program for   .zoo, tried: zoo
Sep 25 23:11:27 mail amavis[684]: No ext program for   .doc, tried: ripole
Sep 25 23:11:27 mail amavis[684]: No decoder for       .F   
Sep 25 23:11:27 mail amavis[684]: No decoder for       .doc 
Sep 25 23:11:27 mail amavis[684]: No decoder for       .zoo 
Sep 25 23:11:27 mail amavis[684]: Using primary internal av scanner code for ClamAV-clamd
Sep 25 23:11:27 mail amavis[684]: Found secondary av scanner ClamAV-clamscan at /usr/bin/clamscan
Sep 25 23:11:57 mail postfix/smtp[682]: connect to gmail-smtp-in.l.google.com[173.194.76.27]:25: Connection timed out
Sep 25 23:12:27 mail postfix/smtp[682]: connect to alt1.gmail-smtp-in.l.google.com[142.250.153.27]:25: Connection timed out
Sep 25 23:12:27 mail postfix/smtp[682]: connect to alt1.gmail-smtp-in.l.google.com[2a00:1450:4013:c16::1a]:25: Cannot assign requested address
Sep 25 23:12:27 mail postfix/smtp[682]: connect to alt2.gmail-smtp-in.l.google.com[2a00:1450:4025:c03::1a]:25: Cannot assign requested address
Sep 25 23:12:27 mail postfix/smtp[682]: 83A649201BD: to=<example.thibaut@gmail.com>, relay=none, delay=6265, delays=6205/0.02/60/0, dsn=4.4.1, status=deferred (connect to alt2.gmail-smtp-in.l.google.com[2a00:1450:4025:c03::1a]:25: Cannot assign requested address)
```

**Comment by polarathene** [comment]
You'll find a recent issue with Traefik is detailed here too.

> submissions/inet/smtpd_upstream_proxy_protocol=haproxy

This is correct for upcoming v13 (`:edge`), you're running 12.1. You have not viewed the correct docs (_they are versioned by image tag_).

I'd like to know how this is happening? Our docs should link you to `:latest` by default, and you'll see [the correct service name in `v12.1` is `smtps` not `submissions`](https://docker-mailserver.github.io/docker-mailserver/v12.1/examples/tutorials/mailserver-behind-proxy/#configuration-of-the-backend-dovecot-and-postfix).

I take it you referenced config from an existing issue? Perhaps https://github.com/docker-mailserver/docker-mailserver/issues/3546

**Comment by sanjuant** [comment]
Thank you very much for your feedback. I would have kept looking for a long time.

To find this configuration, I googled "mailserver traefik". And I found the page https://docker-mailserver.github.io/docker-mailserver/edge/examples/tutorials/mailserver-behind-proxy/

Unfortunately, this is the first time I've seen the EDGE tag, and I must admit that the site header misled me.

![image](https://github.com/docker-mailserver/docker-mailserver/assets/27998877/c7c5bb44-31ce-4bad-8c14-1e0fcaf2c963)

The tag disappears after scrolling, I think that's why I didn't pay attention to it.

Now I have a timeout error, I can continue debugging. Thanks a lot.

**Comment by sanjuant** [comment]
I have a question, I managed to configure the certificate. I can send messages from my gmail mailbox to my dms server, but I can't connect to Thunderbird. What ports do I need to configure?

![image](https://github.com/docker-mailserver/docker-mailserver/assets/27998877/6c77379e-29f6-488b-84a2-510f17173904)
<details>
  <summary>logs</summary>
  
```
[   INF   ]  Welcome to docker-mailserver 12.1.0
[   INF   ]  Checking configuration
[   INF   ]  Configuring mail server
[ WARNING ]  (Rspamd setup) Running Amavis/SA & Rspamd at the same time is discouraged
[ WARNING ]  (Rspamd setup) Running OpenDMARC & Rspamd at the same time is discouraged - we recommend Rspamd for DMARC checks (enabled with Rspamd by default)
[ WARNING ]  (Rspamd setup) Running policyd-spf & Rspamd at the same time is discouraged - we recommend Rspamd for SPF checks (enabled with Rspamd by default)
[   INF   ]  Starting daemons
[   INF   ]  mail.sanjuan.pro is up and running
Sep 27 20:26:35 mail amavis[666]: starting. /usr/sbin/amavisd-new at mail.sanjuan.pro amavisd-new-2.11.1 (20181009), Unicode aware, LC_CTYPE="C.UTF-8"
Sep 27 20:26:35 mail amavis[666]: perl=5.032001, user=, EUID: 109 (109);  group=, EGID: 111 111 (111 111)
Sep 27 20:26:35 mail amavis[666]: Net::Server: Group Not Defined.  Defaulting to EGID '111 111'
Sep 27 20:26:35 mail amavis[666]: Net::Server: User Not Defined.  Defaulting to EUID '109'
Sep 27 20:26:35 mail amavis[666]: No ext program for   .zoo, tried: zoo
Sep 27 20:26:35 mail amavis[666]: No ext program for   .doc, tried: ripole
Sep 27 20:26:35 mail amavis[666]: No decoder for       .F   
Sep 27 20:26:35 mail amavis[666]: No decoder for       .doc 
Sep 27 20:26:35 mail amavis[666]: No decoder for       .zoo 
Sep 27 20:26:35 mail amavis[666]: Using primary internal av scanner code for ClamAV-clamd
Sep 27 20:26:35 mail amavis[666]: Found secondary av scanner ClamAV-clamscan at /usr/bin/clamscan
Sep 27 20:26:55 mail postfix/postscreen[748]: CONNECT from [167.71.160.115]:40520 to [172.21.0.2]:25
Sep 27 20:26:55 mail postfix/postscreen[748]: PASS OLD [167.71.160.115]:40520
Sep 27 20:26:55 mail postfix/smtpd[749]: connect from www11-do.checktls.com[167.71.160.115]
Sep 27 20:26:56 mail postfix/smtpd[749]: Anonymous TLS connection established from www11-do.checktls.com[167.71.160.115]: TLSv1.3 with cipher TLS_AES_256_GCM_SHA384 (256/256 bits) key-exchange X25519 server-signature RSA-PSS (4096 bits) server-digest SHA256
Sep 27 20:26:56 mail postfix/smtpd[749]: disconnect from www11-do.checktls.com[167.71.160.115] ehlo=2 starttls=1 mail=1 quit=1 commands=5
Sep 27 20:29:36 mail postfix/postscreen[994]: CONNECT from [209.85.222.51]:50569 to [172.21.0.2]:25
Sep 27 20:29:42 mail postfix/postscreen[994]: PASS NEW [209.85.222.51]:50569
Sep 27 20:29:43 mail postfix/smtpd[1004]: connect from mail-ua1-f51.google.com[209.85.222.51]
Sep 27 20:29:43 mail postfix/smtpd[1004]: Anonymous TLS connection established from mail-ua1-f51.google.com[209.85.222.51]: TLSv1.3 with cipher TLS_AES_256_GCM_SHA384 (256/256 bits) key-exchange X25519 server-signature RSA-PSS (4096 bits) server-digest SHA256
Sep 27 20:29:43 mail policyd-spf[1008]: prepend Received-SPF: Pass (mailfrom) identity=mailfrom; client-ip=209.85.222.51; helo=mail-ua1-f51.google.com; envelope-from=sanjuan.thibaut@gmail.com; receiver=<UNKNOWN> 
Sep 27 20:29:43 mail postfix/smtpd[1004]: DC024920156: client=mail-ua1-f51.google.com[209.85.222.51]
Sep 27 20:29:43 mail postfix/cleanup[1011]: DC024920156: message-id=<CAAQa6VJKSjAWjUuOHdc=Nj5CHSPemVXiDZ8MJJVZjD+5G6FYbg@mail.gmail.com>
Sep 27 20:29:43 mail opendmarc[540]: DC024920156: gmail.com pass
Sep 27 20:29:44 mail postfix/qmgr[665]: DC024920156: from=<sanjuan.thibaut@gmail.com>, size=3617, nrcpt=1 (queue active)
Sep 27 20:29:44 mail postfix/smtpd-amavis/smtpd[1018]: connect from localhost[127.0.0.1]
Sep 27 20:29:44 mail postfix/smtpd-amavis/smtpd[1018]: 87C8092017E: client=localhost[127.0.0.1]
Sep 27 20:29:44 mail postfix/cleanup[1011]: 87C8092017E: message-id=<CAAQa6VJKSjAWjUuOHdc=Nj5CHSPemVXiDZ8MJJVZjD+5G6FYbg@mail.gmail.com>
Sep 27 20:29:44 mail postfix/qmgr[665]: 87C8092017E: from=<sanjuan.thibaut@gmail.com>, size=4781, nrcpt=1 (queue active)
Sep 27 20:29:44 mail amavis[715]: (00715-01) Passed CLEAN {RelayedInbound}, [209.85.222.51]:50569 [209.85.222.51] <sanjuan.thibaut@gmail.com> -> <admin@sanjuan.pro>, Queue-ID: DC024920156, Message-ID: <CAAQa6VJKSjAWjUuOHdc=Nj5CHSPemVXiDZ8MJJVZjD+5G6FYbg@mail.gmail.com>, mail_id: xjBigYrBgb8A, Hits: -, size: 4584, queued_as: 87C8092017E, 85 ms
Sep 27 20:29:44 mail postfix/smtp-amavis/smtp[1016]: DC024920156: to=<admin@sanjuan.pro>, relay=127.0.0.1[127.0.0.1]:10024, delay=0.89, delays=0.8/0.01/0/0.08, dsn=2.0.0, status=sent (250 2.0.0 from MTA(smtp:[127.0.0.1]:10025): 250 2.0.0 Ok: queued as 87C8092017E)
Sep 27 20:29:44 mail postfix/qmgr[665]: DC024920156: removed
Sep 27 20:29:44 mail dovecot: lmtp(1020): Connect from local
Sep 27 20:29:44 mail postfix/smtpd[1004]: disconnect from mail-ua1-f51.google.com[209.85.222.51] ehlo=2 starttls=1 mail=1 rcpt=1 bdat=1 quit=1 commands=7
Sep 27 20:29:44 mail dovecot: lmtp(admin@sanjuan.pro)<1020><wb0SIph0FGX8AwAA1UCgnw>: sieve: msgid=<CAAQa6VJKSjAWjUuOHdc=Nj5CHSPemVXiDZ8MJJVZjD+5G6FYbg@mail.gmail.com>: stored mail into mailbox 'INBOX'
Sep 27 20:29:44 mail postfix/lmtp[1019]: 87C8092017E: to=<admin@sanjuan.pro>, relay=mail.sanjuan.pro[/var/run/dovecot/lmtp], delay=0.02, delays=0/0.01/0.01/0.01, dsn=2.0.0, status=sent (250 2.0.0 <admin@sanjuan.pro> wb0SIph0FGX8AwAA1UCgnw Saved)
Sep 27 20:29:44 mail dovecot: lmtp(1020): Disconnect from local: Logged out (state=READY)
Sep 27 20:29:44 mail postfix/qmgr[665]: 87C8092017E: removed
```
  
</details>

![image](https://github.com/docker-mailserver/docker-mailserver/assets/27998877/ac535d73-fb76-4ade-8234-821acc92fb18)![image](https://github.com/docker-mailserver/docker-mailserver/assets/27998877/d4d3c401-bd4d-4c86-a10a-52487dc398e4)

**Comment by polarathene** [comment]
Does this help?: https://github.com/orgs/docker-mailserver/discussions/2575#discussioncomment-3439509

**Comment by sanjuant** [comment]
Thanks for the link, but it didn't help. I've been trying to set up DMS for almost 15 days, I'm going to give up x). I've started all over again, 

traefik compose
```yaml
services:
  dockerproxy: # see https://github.com/Tecnativa/docker-socket-proxy
    container_name: dockerproxy
    image: ghcr.io/tecnativa/docker-socket-proxy:0.1.1 # this image is rather old but used to have a pinned version
                                                       # newer version is ghcr.io/tecnativa/docker-socket-proxy:edge
    restart: unless-stopped
    mem_limit: 2G
    cpus: 0.75
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock:ro   # never expose this container to the internet!
    environment:
      CONTAINERS: 1
    networks:
      - docker-proxynet # use only internal network

  traefik:
    container_name: traefik
    image: traefik:2.8 # use current version 2.8.x
    restart: unless-stopped
    read_only: true
    mem_limit: 2G
    cpus: 0.75
    depends_on:
      - dockerproxy
    security_opt:
      - no-new-privileges:true
    volumes:
      - /data/traefik/config/traefik.yaml:/etc/traefik/traefik.yaml:ro    # Static Traefik Configuration
      - /data/traefik/config/dynamic/:/etc/traefik/dynamic/:ro            # Folder to store dynamic configuration file provider
      - /data/traefik/config/authusers.txt:/etc/traefik/authusers.txt:ro  # userfile for basic auth
      - /data/traefik/config/acme/acme.json:/etc/traefik/acme.json        # certificate storage
    user: "2000:2000" # user traefik has to be created on the host system
                      # sudo useradd -u 2000 -M -s /usr/sbin/nologin traefik
    labels:
      - "traefik.enable=true"
      # expose traefik dashboard with TLS and basic auth
      # 1. create router
      - "traefik.http.routers.traefik.rule=Host(`traefik.sanjuan.pro`)"
      - "traefik.http.routers.traefik.service=api@internal"
      - "traefik.http.routers.traefik.tls.certresolver=tlschallenge"
      - "traefik.http.routers.traefik.entrypoints=web-secure"
      - "traefik.http.routers.traefik.middlewares=traefik-auth, secHeaders@file"
      # 2. create basic auth middleware
      - "traefik.http.middlewares.traefik-auth.basicauth.removeheader=true"
      - "traefik.http.middlewares.traefik-auth.basicauth.usersfile=/etc/traefik/authusers.txt"
      # this would be another possibility to create an catch-all on port 80 with redirection to https
      # but we connected the unencrypted entrypoint to the redirection in the static configuration:
      #- "traefik.http.routers.http-catchall.rule=hostregexp(`{host:.+}`)"
      #- "traefik.http.routers.http-catchall.entrypoints=web"
      #- "traefik.http.routers.http-catchall.middlewares=redirect-to-https"
      #- "traefik.http.middlewares.redirect-to-https.redirectscheme.scheme=https"
    networks:
      - traefik-servicenet
      - docker-proxynet
    ports:
      - "80:10080"  # use high ports inside the container so
      - "443:10443" # we don't need to be root to bind the ports
      - "25:25"
      - "465:465"
      - "993:993"
      - "4190:4190"

networks:
  traefik-servicenet:
    external: true           # this network has to be created once before starting:
    name: traefik-servicenet # docker network create traefik-servicenet
  docker-proxynet:
    driver: bridge
    internal: true
```

dms compose
```yaml
services:
  mailserver:
    image: ghcr.io/docker-mailserver/docker-mailserver:latest
    container_name: mailserver
    # Provide the FQDN of your mail server here (Your DNS MX record should point to this value)
    hostname: mail.sanjuan.pro
    #domainname: 
    # ports:
    #   - "25:25"
    #   - "465:465"
    #   - "587:587"
    #   - "993:993"
    volumes:
      - /data/dms/mail-data/:/var/mail/
      - /data/dms/mail-state/:/var/mail-state/
      - /data/dms/mail-logs/:/var/log/mail/
      - /data/dms/config/:/tmp/docker-mailserver/
      - /etc/localtime:/etc/localtime:ro
      # - /data/traefik/config/acme/acme.json:/etc/letsencrypt/acme.json:ro
    environment:
      # - SSL_TYPE=letsencrypt
      # - SSL_DOMAIN=mail.sanjuan.pro
      - ENABLE_OPENDKIM=0
      - ENABLE_RSPAMD=1
      - ENABLE_CLAMAV=1
      - ENABLE_FAIL2BAN=0
    cap_add:
      - NET_ADMIN # For Fail2Ban to work
    labels:
      - "traefik.enable=true"
      - "traefik.tcp.routers.smtp.rule=HostSNI(`*`)"
      - "traefik.tcp.routers.smtp.entrypoints=smtp"
      - "traefik.tcp.routers.smtp.service=smtp"
      - "traefik.tcp.services.smtp.loadbalancer.server.port=25"
      - "traefik.tcp.services.smtp.loadbalancer.proxyProtocol.version=1"
      - "traefik.tcp.routers.smtp-ssl.rule=HostSNI(`*`)"
      - "traefik.tcp.routers.smtp-ssl.tls=false"
      - "traefik.tcp.routers.smtp-ssl.entrypoints=smtp-ssl"
      - "traefik.tcp.routers.smtp-ssl.service=smtp-ssl"
      - "traefik.tcp.services.smtp-ssl.loadbalancer.server.port=465"
      - "traefik.tcp.services.smtp-ssl.loadbalancer.proxyProtocol.version=1"
      - "traefik.tcp.routers.imap-ssl.rule=HostSNI(`*`)"
      - "traefik.tcp.routers.imap-ssl.entrypoints=imap-ssl"
      - "traefik.tcp.routers.imap-ssl.service=imap-ssl"
      - "traefik.tcp.services.imap-ssl.loadbalancer.server.port=10993"
      - "traefik.tcp.services.imap-ssl.loadbalancer.proxyProtocol.version=2"
      - "traefik.tcp.routers.sieve.rule=HostSNI(`*`)"
      - "traefik.tcp.routers.sieve.entrypoints=sieve"
      - "traefik.tcp.routers.sieve.service=sieve"
      - "traefik.tcp.services.sieve.loadbalancer.server.port=4190"
    restart: always
    networks:
      - traefik-servicenet

networks:
  traefik-servicenet:
    external: true
```

dovecot-quotas.cf
```
haproxy_trusted_networks = 172.21.0.2 # traefik-servicenet
haproxy_timeout = 3 sec
service imap-login {
  inet_listener imaps {
    haproxy = yes
    ssl = yes
    port = 10993
  }
}
```

post-main.cf
```
postscreen_upstream_proxy_protocol = haproxy
```

post-master.cf
```
submission/inet/smtpd_upstream_proxy_protocol=haproxy
smtps/inet/smtpd_upstream_proxy_protocol=haproxy
```

At this level, I'm able to send e-mails 
- from inside to inside
- from outside to inside 

but I can't send mail 
- from inside to outside.

<details>
  <summary>logs</summary>
  
```
[   INF   ]  Welcome to docker-mailserver 12.1.0
[   INF   ]  Checking configuration
[   INF   ]  Configuring mail server
[ WARNING ]  (Rspamd setup) Running Amavis/SA & Rspamd at the same time is discouraged
[ WARNING ]  (Rspamd setup) Running OpenDMARC & Rspamd at the same time is discouraged - we recommend Rspamd for DMARC checks (enabled with Rspamd by default)
[ WARNING ]  (Rspamd setup) Running policyd-spf & Rspamd at the same time is discouraged - we recommend Rspamd for SPF checks (enabled with Rspamd by default)
[ WARNING ]  !! INSECURE !! SSL configured with plain text access - DO NOT USE FOR PRODUCTION DEPLOYMENT
[   INF   ]  Starting daemons
[   INF   ]  mail.sanjuan.pro is up and running
Sep 28 22:07:01 mail amavis[617]: starting. /usr/sbin/amavisd-new at mail.sanjuan.pro amavisd-new-2.11.1 (20181009), Unicode aware, LC_CTYPE="C.UTF-8"
Sep 28 22:07:01 mail amavis[617]: perl=5.032001, user=, EUID: 109 (109);  group=, EGID: 111 111 (111 111)
Sep 28 22:07:01 mail amavis[617]: Net::Server: Group Not Defined.  Defaulting to EGID '111 111'
Sep 28 22:07:01 mail amavis[617]: Net::Server: User Not Defined.  Defaulting to EUID '109'
Sep 28 22:07:01 mail amavis[617]: No ext program for   .zoo, tried: zoo
Sep 28 22:07:01 mail amavis[617]: No ext program for   .doc, tried: ripole
Sep 28 22:07:01 mail amavis[617]: No decoder for       .F   
Sep 28 22:07:01 mail amavis[617]: No decoder for       .doc 
Sep 28 22:07:01 mail amavis[617]: No decoder for       .zoo 
Sep 28 22:07:01 mail amavis[617]: Using primary internal av scanner code for ClamAV-clamd
Sep 28 22:07:01 mail amavis[617]: Found secondary av scanner ClamAV-clamscan at /usr/bin/clamscan
Sep 28 22:07:09 mail postfix/postscreen[709]: CONNECT from [172.21.0.4]:46050 to [172.21.0.2]:25
Sep 28 22:07:09 mail postfix/postscreen[709]: PASS OLD [172.21.0.4]:46050
Sep 28 22:07:09 mail postfix/smtpd[710]: connect from jfa.traefik-servicenet[172.21.0.4]
Sep 28 22:07:09 mail postfix/trivial-rewrite[713]: warning: do not list domain mail.sanjuan.pro in BOTH mydestination and virtual_mailbox_domains
Sep 28 22:07:09 mail postfix/trivial-rewrite[713]: warning: do not list domain mail.sanjuan.pro in BOTH mydestination and virtual_mailbox_domains
Sep 28 22:07:09 mail policyd-spf[715]: prepend Received-SPF: None (mailfrom) identity=mailfrom; client-ip=172.21.0.4; helo=mail.sanjuan.pro; envelope-from=admin@mail.sanjuan.pro; receiver=<UNKNOWN> 
Sep 28 22:07:09 mail postfix/smtpd[710]: C20B3920157: client=jfa.traefik-servicenet[172.21.0.4]
Sep 28 22:07:09 mail postfix/cleanup[718]: C20B3920157: message-id=<>
Sep 28 22:07:09 mail opendmarc[540]: C20B3920157: mail.sanjuan.pro none
Sep 28 22:07:10 mail postfix/qmgr[649]: C20B3920157: from=<admin@mail.sanjuan.pro>, size=10323, nrcpt=1 (queue active)
Sep 28 22:07:10 mail postfix/smtpd[710]: disconnect from jfa.traefik-servicenet[172.21.0.4] ehlo=1 mail=1 rcpt=1 data=1 quit=1 commands=5
Sep 28 22:07:10 mail amavis[695]: (00695-01) (!)connect to /var/run/clamav/clamd.ctl failed, attempt #1: Can't connect to a UNIX socket /var/run/clamav/clamd.ctl: No such file or directory
Sep 28 22:07:11 mail amavis[695]: (00695-01) (!)connect to /var/run/clamav/clamd.ctl failed, attempt #1: Can't connect to a UNIX socket /var/run/clamav/clamd.ctl: No such file or directory
Sep 28 22:07:11 mail amavis[695]: (00695-01) (!)ClamAV-clamd: All attempts (1) failed connecting to /var/run/clamav/clamd.ctl, retrying (2)
Sep 28 22:07:17 mail postfix/smtpd-amavis/smtpd[732]: connect from localhost[127.0.0.1]
Sep 28 22:07:17 mail postfix/trivial-rewrite[713]: warning: do not list domain mail.sanjuan.pro in BOTH mydestination and virtual_mailbox_domains
Sep 28 22:07:17 mail postfix/smtpd-amavis/smtpd[732]: 335319201BD: client=localhost[127.0.0.1]
Sep 28 22:07:17 mail postfix/cleanup[718]: 335319201BD: message-id=<20230928200717.335319201BD@mail.sanjuan.pro>
Sep 28 22:07:17 mail postfix/qmgr[649]: 335319201BD: from=<admin@mail.sanjuan.pro>, size=10642, nrcpt=1 (queue active)
Sep 28 22:07:17 mail amavis[695]: (00695-01) Passed CLEAN {RelayedInbound}, [172.21.0.4]:46050 <admin@mail.sanjuan.pro> -> <admin@sanjuan.pro>, Queue-ID: C20B3920157, mail_id: wlLZlP6ZyBdf, Hits: -, size: 10386, queued_as: 335319201BD, 7094 ms
Sep 28 22:07:17 mail postfix/smtp-amavis/smtp[719]: C20B3920157: to=<admin@sanjuan.pro>, relay=127.0.0.1[127.0.0.1]:10024, delay=7.5, delays=0.4/0.01/0/7.1, dsn=2.0.0, status=sent (250 2.0.0 from MTA(smtp:[127.0.0.1]:10025): 250 2.0.0 Ok: queued as 335319201BD)
Sep 28 22:07:17 mail postfix/qmgr[649]: C20B3920157: removed
Sep 28 22:07:17 mail dovecot: lmtp(737): Connect from local
Sep 28 22:07:17 mail dovecot: lmtp(admin@sanjuan.pro)<737><VcOYDfXcFWXhAgAA1UCgnw>: sieve: msgid=<20230928200717.335319201BD@mail.sanjuan.pro>: stored mail into mailbox 'INBOX'
Sep 28 22:07:17 mail dovecot: lmtp(737): Disconnect from local: Logged out (state=READY)
Sep 28 22:07:17 mail postfix/lmtp[733]: 335319201BD: to=<admin@sanjuan.pro>, relay=mail.sanjuan.pro[/var/run/dovecot/lmtp], delay=0.03, delays=0/0.01/0.01/0.01, dsn=2.0.0, status=sent (250 2.0.0 <admin@sanjuan.pro> VcOYDfXcFWXhAgAA1UCgnw Saved)
Sep 28 22:07:17 mail postfix/qmgr[649]: 335319201BD: removed
Sep 28 22:08:05 mail postfix/postscreen[709]: CONNECT from [209.85.221.170]:45387 to [172.21.0.2]:25
Sep 28 22:08:11 mail postfix/postscreen[709]: PASS NEW [209.85.221.170]:45387
Sep 28 22:08:11 mail postfix/smtpd[710]: connect from mail-vk1-f170.google.com[209.85.221.170]
Sep 28 22:08:12 mail policyd-spf[715]: prepend Received-SPF: Pass (mailfrom) identity=mailfrom; client-ip=209.85.221.170; helo=mail-vk1-f170.google.com; envelope-from=seekcoldgirl@gmail.com; receiver=<UNKNOWN> 
Sep 28 22:08:12 mail postfix/smtpd[710]: 489B99201AB: client=mail-vk1-f170.google.com[209.85.221.170]
Sep 28 22:08:12 mail postfix/cleanup[718]: 489B99201AB: message-id=<CADR1n6-6bJ12RFzuAwZ_OjYQyvFHnMBcDALKFXqy-kjC-4o0ng@mail.gmail.com>
Sep 28 22:08:12 mail opendmarc[540]: 489B99201AB: gmail.com pass
Sep 28 22:08:12 mail postfix/qmgr[649]: 489B99201AB: from=<seekcoldgirl@gmail.com>, size=2960, nrcpt=1 (queue active)
Sep 28 22:08:12 mail postfix/smtpd-amavis/smtpd[821]: connect from localhost[127.0.0.1]
Sep 28 22:08:12 mail postfix/smtpd-amavis/smtpd[821]: B52EF9201C0: client=localhost[127.0.0.1]
Sep 28 22:08:12 mail postfix/cleanup[718]: B52EF9201C0: message-id=<CADR1n6-6bJ12RFzuAwZ_OjYQyvFHnMBcDALKFXqy-kjC-4o0ng@mail.gmail.com>
Sep 28 22:08:12 mail postfix/qmgr[649]: B52EF9201C0: from=<seekcoldgirl@gmail.com>, size=4214, nrcpt=1 (queue active)
Sep 28 22:08:12 mail dovecot: lmtp(737): Connect from local
Sep 28 22:08:12 mail amavis[696]: (00696-01) Passed CLEAN {RelayedInbound}, [209.85.221.170]:45387 [209.85.221.170] <seekcoldgirl@gmail.com> -> <admin@sanjuan.pro>, Queue-ID: 489B99201AB, Message-ID: <CADR1n6-6bJ12RFzuAwZ_OjYQyvFHnMBcDALKFXqy-kjC-4o0ng@mail.gmail.com>, mail_id: i4B53aebNagQ, Hits: -, size: 4017, queued_as: B52EF9201C0, 78 ms
Sep 28 22:08:12 mail postfix/smtp-amavis/smtp[719]: 489B99201AB: to=<admin@sanjuan.pro>, relay=127.0.0.1[127.0.0.1]:10024, delay=0.59, delays=0.51/0/0/0.08, dsn=2.0.0, status=sent (250 2.0.0 from MTA(smtp:[127.0.0.1]:10025): 250 2.0.0 Ok: queued as B52EF9201C0)
Sep 28 22:08:12 mail postfix/qmgr[649]: 489B99201AB: removed
Sep 28 22:08:12 mail dovecot: lmtp(admin@sanjuan.pro)<737><sNFqLCzdFWXhAgAA1UCgnw>: sieve: msgid=<CADR1n6-6bJ12RFzuAwZ_OjYQyvFHnMBcDALKFXqy-kjC-4o0ng@mail.gmail.com>: stored mail into mailbox 'INBOX'
Sep 28 22:08:12 mail postfix/lmtp[733]: B52EF9201C0: to=<admin@sanjuan.pro>, relay=mail.sanjuan.pro[/var/run/dovecot/lmtp], delay=0.01, delays=0/0/0/0.01, dsn=2.0.0, status=sent (250 2.0.0 <admin@sanjuan.pro> sNFqLCzdFWXhAgAA1UCgnw Saved)
Sep 28 22:08:12 mail dovecot: lmtp(737): Disconnect from local: Logged out (state=READY)
Sep 28 22:08:12 mail postfix/qmgr[649]: B52EF9201C0: removed
Sep 28 22:08:12 mail postfix/smtpd[710]: disconnect from mail-vk1-f170.google.com[209.85.221.170] ehlo=1 mail=1 rcpt=1 bdat=1 quit=1 commands=5
Sep 28 22:08:45 mail postfix/postscreen[709]: CONNECT from [172.21.0.4]:51516 to [172.21.0.2]:25
Sep 28 22:08:45 mail postfix/postscreen[709]: PASS OLD [172.21.0.4]:51516
Sep 28 22:08:45 mail postfix/smtpd[710]: connect from jfa.traefik-servicenet[172.21.0.4]
Sep 28 22:08:45 mail postfix/smtpd[710]: NOQUEUE: reject: RCPT from jfa.traefik-servicenet[172.21.0.4]: 554 5.7.1 <seekcoldgirl@gmail.com>: Relay access denied; from=<admin@mail.sanjuan.pro> to=<seekcoldgirl@gmail.com> proto=ESMTP helo=<mail.sanjuan.pro>
Sep 28 22:08:45 mail postfix/smtpd[710]: disconnect from jfa.traefik-servicenet[172.21.0.4] ehlo=1 mail=1 rcpt=0/1 quit=1 commands=3/4
```
  
</details>

No connection problems
![image](https://github.com/docker-mailserver/docker-mailserver/assets/27998877/e5e04ea9-adcc-4046-84d8-0c814ab71ac7)

However, having commented on the following, the checktls don't work.
```yaml
      # - /data/traefik/config/acme/acme.json:/etc/letsencrypt/acme.json:ro
    environment:
      # - SSL_TYPE=letsencrypt
      # - SSL_DOMAIN=mail.sanjuan.pro
```
![image](https://github.com/docker-mailserver/docker-mailserver/assets/27998877/16f2a03b-25c6-441d-854a-3bf5a04b2db0)

**Comment by polarathene** [comment]
> Thanks for the link, but it didn't help. I've been trying to set up DMS for almost 15 days, I'm going to give up x)

Can you get this to work without traefik + haproxy involved? You can still use the `acme.json` with `SSL_TYPE=letsencrypt` but I'm not sure if traefik is benefiting you much since only DMS should need its own ports?

Just curious if [like this other Traefik compatibility report](https://github.com/docker-mailserver/docker-mailserver/issues/3546#issuecomment-1738979539), that something is wrong with traefik integration only for you?

**Comment by sanjuant** [comment]
Without using traefik, I can connect with Thundebird, but I can't send an e-mail to gmails. In addition, for security reasons, I'd like to expose only the Traefik container on the Internet, and route ports through it.

![image](https://github.com/docker-mailserver/docker-mailserver/assets/27998877/be784bd8-31df-4ce3-b2b1-690f48402f0c)
![image](https://github.com/docker-mailserver/docker-mailserver/assets/27998877/2f097a3b-16b4-4e2b-b093-ea81475d3d14)

<details>
  <summary>logs</summary>
  
```
Sep 29 03:02:33 mail postfix/anvil[818]: statistics: max connection count 4 for (submission:83.113.65.231) at Sep 29 02:53:04

Sep 29 03:02:33 mail postfix/anvil[818]: statistics: max cache size 3 at Sep 29 02:53:04

Sep 29 03:02:35 mail dovecot: imap(admin@sanjuan.pro)<1060><mRPm33QGetdTcUHn>: Disconnected: Logged out in=814 out=44146 deleted=0 expunged=0 trashed=0 hdr_count=5 hdr_bytes=1276 body_count=5 body_bytes=39353

Sep 29 03:02:49 mail postfix/postscreen[1777]: CONNECT from [83.113.65.231]:55305 to [172.21.0.7]:25

Sep 29 03:02:49 mail postfix/postscreen[1777]: CONNECT from [83.113.65.231]:55307 to [172.21.0.7]:25

Sep 29 03:02:49 mail postfix/postscreen[1777]: CONNECT from [83.113.65.231]:55309 to [172.21.0.7]:25

Sep 29 03:02:49 mail postfix/postscreen[1777]: CONNECT from [83.113.65.231]:55311 to [172.21.0.7]:25

Sep 29 03:02:49 mail postfix/postscreen[1777]: PREGREET 33 after 0 from [83.113.65.231]:55305: EHLO we-guess.mozilla.org\r\nQUIT\r\n

Sep 29 03:02:49 mail postfix/postscreen[1777]: PREGREET 33 after 0 from [83.113.65.231]:55309: EHLO we-guess.mozilla.org\r\nQUIT\r\n

Sep 29 03:02:49 mail postfix/postscreen[1777]: COMMAND PIPELINING from [83.113.65.231]:55305 after EHLO: QUIT\r\n

Sep 29 03:02:49 mail postfix/postscreen[1777]: DISCONNECT [83.113.65.231]:55305

Sep 29 03:02:49 mail postfix/postscreen[1777]: COMMAND PIPELINING from [83.113.65.231]:55309 after EHLO: QUIT\r\n

Sep 29 03:02:49 mail postfix/postscreen[1777]: DISCONNECT [83.113.65.231]:55309

Sep 29 03:02:49 mail postfix/postscreen[1777]: PREGREET 33 after 0.01 from [83.113.65.231]:55307: EHLO we-guess.mozilla.org\r\nQUIT\r\n

Sep 29 03:02:49 mail postfix/postscreen[1777]: COMMAND PIPELINING from [83.113.65.231]:55307 after EHLO: QUIT\r\n

Sep 29 03:02:49 mail postfix/postscreen[1777]: DISCONNECT [83.113.65.231]:55307

Sep 29 03:02:49 mail postfix/postscreen[1777]: PREGREET 33 after 0.01 from [83.113.65.231]:55311: EHLO we-guess.mozilla.org\r\nQUIT\r\n

Sep 29 03:02:49 mail postfix/submission/smtpd[1776]: connect from atoulon-651-1-170-231.w83-113.abo.wanadoo.fr[83.113.65.231]

Sep 29 03:02:49 mail postfix/postscreen[1777]: COMMAND PIPELINING from [83.113.65.231]:55311 after EHLO: QUIT\r\n

Sep 29 03:02:49 mail postfix/postscreen[1777]: DISCONNECT [83.113.65.231]:55311

Sep 29 03:02:49 mail postfix/smtps/smtpd[1775]: connect from atoulon-651-1-170-231.w83-113.abo.wanadoo.fr[83.113.65.231]

Sep 29 03:02:49 mail postfix/submission/smtpd[1781]: connect from atoulon-651-1-170-231.w83-113.abo.wanadoo.fr[83.113.65.231]

Sep 29 03:02:49 mail postfix/smtps/smtpd[1782]: connect from atoulon-651-1-170-231.w83-113.abo.wanadoo.fr[83.113.65.231]

Sep 29 03:02:49 mail postfix/submission/smtpd[1783]: connect from atoulon-651-1-170-231.w83-113.abo.wanadoo.fr[83.113.65.231]

Sep 29 03:02:49 mail postfix/submission/smtpd[1776]: improper command pipelining after EHLO from atoulon-651-1-170-231.w83-113.abo.wanadoo.fr[83.113.65.231]: QUIT\r\n

Sep 29 03:02:49 mail postfix/submission/smtpd[1781]: improper command pipelining after EHLO from atoulon-651-1-170-231.w83-113.abo.wanadoo.fr[83.113.65.231]: QUIT\r\n

Sep 29 03:02:49 mail postfix/smtps/smtpd[1775]: Anonymous TLS connection established from atoulon-651-1-170-231.w83-113.abo.wanadoo.fr[83.113.65.231]: TLSv1.3 with cipher TLS_AES_256_GCM_SHA384 (256/256 bits) key-exchange X25519 server-signature RSA-PSS (4096 bits)
```
  
</details>

traefik compose
```yaml
services:
  dockerproxy: # see https://github.com/Tecnativa/docker-socket-proxy
    container_name: dockerproxy
    image: ghcr.io/tecnativa/docker-socket-proxy:0.1.1 # this image is rather old but used to have a pinned version
                                                       # newer version is ghcr.io/tecnativa/docker-socket-proxy:edge
    restart: unless-stopped
    mem_limit: 2G
    cpus: 0.75
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock:ro   # never expose this container to the internet!
    environment:
      CONTAINERS: 1
    networks:
      - docker-proxynet # use only internal network

  traefik:
    container_name: traefik
    image: traefik:2.8 # use current version 2.8.x
    restart: unless-stopped
    read_only: true
    mem_limit: 2G
    cpus: 0.75
    depends_on:
      - dockerproxy
    security_opt:
      - no-new-privileges:true
    volumes:
      - /data/traefik/config/traefik.yaml:/etc/traefik/traefik.yaml:ro    # Static Traefik Configuration
      - /data/traefik/config/dynamic/:/etc/traefik/dynamic/:ro            # Folder to store dynamic configuration file provider
      - /data/traefik/config/authusers.txt:/etc/traefik/authusers.txt:ro  # userfile for basic auth
      - /data/traefik/config/acme/acme.json:/etc/traefik/acme.json        # certificate storage
    user: "2000:2000" # user traefik has to be created on the host system
                      # sudo useradd -u 2000 -M -s /usr/sbin/nologin traefik
    labels:
      - "traefik.enable=true"
      # expose traefik dashboard with TLS and basic auth
      # 1. create router
      - "traefik.http.routers.traefik.rule=Host(`traefik.sanjuan.pro`)"
      - "traefik.http.routers.traefik.service=api@internal"
      - "traefik.http.routers.traefik.tls.certresolver=tlschallenge"
      - "traefik.http.routers.traefik.entrypoints=web-secure"
      - "traefik.http.routers.traefik.middlewares=traefik-auth, secHeaders@file"
      # 2. create basic auth middleware
      - "traefik.http.middlewares.traefik-auth.basicauth.removeheader=true"
      - "traefik.http.middlewares.traefik-auth.basicauth.usersfile=/etc/traefik/authusers.txt"
      # this would be another possibility to create an catch-all on port 80 with redirection to https
      # but we connected the unencrypted entrypoint to the redirection in the static configuration:
      #- "traefik.http.routers.http-catchall.rule=hostregexp(`{host:.+}`)"
      #- "traefik.http.routers.http-catchall.entrypoints=web"
      #- "traefik.http.routers.http-catchall.middlewares=redirect-to-https"
      #- "traefik.http.middlewares.redirect-to-https.redirectscheme.scheme=https"
    networks:
      - traefik-servicenet
      - docker-proxynet
    ports:
      - "80:10080"  # use high ports inside the container so
      - "443:10443" # we don't need to be root to bind the ports
      # - "25:25"
      # - "465:465"
      # - "993:993"
      # - "4190:4190"

networks:
  traefik-servicenet:
    external: true           # this network has to be created once before starting:
    name: traefik-servicenet # docker network create traefik-servicenet
  docker-proxynet:
    driver: bridge
    internal: true
```

dms compose
```yaml
services:
  mailserver:
    image: ghcr.io/docker-mailserver/docker-mailserver:latest
    container_name: mailserver
    # Provide the FQDN of your mail server here (Your DNS MX record should point to this value)
    hostname: mail.sanjuan.pro
    #domainname: 
    ports:
      - "25:25"
      - "465:465"
      - "587:587"
      - "993:993"
    volumes:
      - /data/dms/mail-data/:/var/mail/
      - /data/dms/mail-state/:/var/mail-state/
      - /data/dms/mail-logs/:/var/log/mail/
      - /data/dms/config/:/tmp/docker-mailserver/
      - /etc/localtime:/etc/localtime:ro
      - /data/traefik/config/acme/acme.json:/etc/letsencrypt/acme.json:ro
    environment:
      - SSL_TYPE=letsencrypt
      # - SSL_DOMAIN=mail.sanjuan.pro
      - ENABLE_OPENDKIM=0
      - ENABLE_RSPAMD=1
      - ENABLE_CLAMAV=1
      - ENABLE_FAIL2BAN=0
    cap_add:
      - NET_ADMIN # For Fail2Ban to work
    labels:
      - "traefik.enable=true"
      
      # - "traefik.http.routers.mailserver.rule=Host(`mail.sanjuan.pro`)"
      # - "traefik.http.routers.mailserver.entrypoints=web-secure"
      # - "traefik.http.routers.mailserver.tls.certresolver=tlschallenge"
      
      # - "traefik.tcp.routers.smtp.rule=HostSNI(`*`)"
      # - "traefik.tcp.routers.smtp.entrypoints=smtp"
      # - "traefik.tcp.routers.smtp.service=smtp"
      # - "traefik.tcp.services.smtp.loadbalancer.server.port=25"
      # - "traefik.tcp.services.smtp.loadbalancer.proxyProtocol.version=1"
      # - "traefik.tcp.routers.smtp-ssl.rule=HostSNI(`*`)"
      # - "traefik.tcp.routers.smtp-ssl.tls=false"
      # - "traefik.tcp.routers.smtp-ssl.entrypoints=smtp-ssl"
      # - "traefik.tcp.routers.smtp-ssl.service=smtp-ssl"
      # - "traefik.tcp.services.smtp-ssl.loadbalancer.server.port=465"
      # - "traefik.tcp.services.smtp-ssl.loadbalancer.proxyProtocol.version=1"
      # - "traefik.tcp.routers.imap-ssl.rule=HostSNI(`*`)"
      # - "traefik.tcp.routers.imap-ssl.entrypoints=imap-ssl"
      # - "traefik.tcp.routers.imap-ssl.service=imap-ssl"
      # - "traefik.tcp.services.imap-ssl.loadbalancer.server.port=10993"
      # - "traefik.tcp.services.imap-ssl.loadbalancer.proxyProtocol.version=2"
      # - "traefik.tcp.routers.sieve.rule=HostSNI(`*`)"
      # - "traefik.tcp.routers.sieve.entrypoints=sieve"
      # - "traefik.tcp.routers.sieve.service=sieve"
      # - "traefik.tcp.services.sieve.loadbalancer.server.port=4190"
    restart: always
    networks:
      - traefik-servicenet

networks:
  traefik-servicenet:
    external: true
```

dovecot-quotas.cf
```
#haproxy_trusted_networks = 172.21.0.2
#haproxy_timeout = 3 secs
#service imap-login {
#  inet_listener imaps {
#    haproxy = yes
#    ssl = yes
#    port = 10993
#  }
#}
```

post-main.cf > post-main.cf.save
```
postscreen_upstream_proxy_protocol = haproxy
```

post-master.cf > post-master.cf.save
```
submission/inet/smtpd_upstream_proxy_protocol=haproxy
smtps/inet/smtpd_upstream_proxy_protocol=haproxy
```

![image](https://github.com/docker-mailserver/docker-mailserver/assets/27998877/972dbe1d-b88c-4739-aac1-c842bd397a27)

**Comment by polarathene** [comment]
> Without using traefik, I can connect with Thundebird, but I can't send an e-mail to gmails.

Relevant logs:

```
Sep 29 02:55:02 mail dovecot: imap-login: Login: user=<admin@sanjuan.pro>, method=PLAIN, rip=83.113.65.231, lip=172.21.0.7, mpid=1061, TLS, session=<mRPm33QGeddTcUHn>
Sep 29 02:55:26 mail postfix/submission/smtpd[1098]: connect from atoulon-651-1-170-231.w83-113.abo.wanadoo.fr[83.113.65.231]
Sep 29 02:55:26 mail postfix/submission/smtpd[1098]: Anonymous TLS connection established from atoulon-651-1-170-231.w83-113.abo.wanadoo.fr[83.113.65.231]: TLSv1.3 with cipher TLS_AES_256_GCM_SHA384 (256/256 bits) key-exchange X25519 server-signature RSA-PSS (4096 bits) server-digest SHA256
Sep 29 02:55:31 mail postfix/submission/smtpd[1098]: 74C10920196: client=atoulon-651-1-170-231.w83-113.abo.wanadoo.fr[83.113.65.231], sasl_method=PLAIN, sasl_username=admin@sanjuan.pro
Sep 29 02:55:31 mail postfix/sender-cleanup/cleanup[1106]: 74C10920196: message-id=<24e8f161-904c-4df5-936b-48b4bfcfc066@sanjuan.pro>


Sep 29 02:55:31 mail postfix/sender-cleanup/cleanup[1106]: 74C10920196: replace: header MIME-Version: 1.0 from atoulon-651-1-170-231.w83-113.abo.wanadoo.fr[83.113.65.231]; from=<admin@sanjuan.pro> to=<sanjuan.thibaut@gmail.com> proto=ESMTP helo=<[10.10.100.28]>: MIME-Version: 1.0
Sep 29 02:55:31 mail postfix/qmgr[648]: 74C10920196: from=<admin@sanjuan.pro>, size=359, nrcpt=1 (queue active)
Sep 29 02:55:31 mail dovecot: imap(admin@sanjuan.pro)<1061><mRPm33QGeddTcUHn>: Disconnected: Logged out in=533 out=992 deleted=0 expunged=0 trashed=0 hdr_count=0 hdr_bytes=0 body_count=0 body_bytes=0
Sep 29 02:55:31 mail postfix/smtpd-amavis/smtpd[1110]: connect from localhost[127.0.0.1]
Sep 29 02:55:31 mail postfix/smtpd-amavis/smtpd[1110]: BFBD0920218: client=localhost[127.0.0.1]
Sep 29 02:55:31 mail postfix/cleanup[1111]: BFBD0920218: message-id=<24e8f161-904c-4df5-936b-48b4bfcfc066@sanjuan.pro>
Sep 29 02:55:31 mail postfix/qmgr[648]: BFBD0920218: from=<admin@sanjuan.pro>, size=526, nrcpt=1 (queue active)
Sep 29 02:55:31 mail amavis[694]: (00694-01) Passed CLEAN {RelayedOpenRelay}, [83.113.65.231]:55171 [83.113.65.231] <admin@sanjuan.pro> -> <sanjuan.thibaut@gmail.com>, Queue-ID: 74C10920196, Message-ID: <24e8f161-904c-4df5-936b-48b4bfcfc066@sanjuan.pro>, mail_id: ANIwvbsqYWkG, Hits: -, size: 322, queued_as: BFBD0920218, 86 ms
Sep 29 02:55:31 mail postfix/smtp-amavis/smtp[1108]: 74C10920196: to=<sanjuan.thibaut@gmail.com>, relay=127.0.0.1[127.0.0.1]:10024, delay=0.33, delays=0.23/0.01/0/0.08, dsn=2.0.0, status=sent (250 2.0.0 from MTA(smtp:[127.0.0.1]:10025): 250 2.0.0 Ok: queued as BFBD0920218)
Sep 29 02:55:31 mail postfix/qmgr[648]: 74C10920196: removed


Sep 29 02:55:32 mail postfix/smtp[1112]: connect to gmail-smtp-in.l.google.com[2a00:1450:400c:c0c::1a]:25: Cannot assign requested address
Sep 29 02:56:02 mail postfix/smtp[1112]: connect to gmail-smtp-in.l.google.com[74.125.71.26]:25: Connection timed out
Sep 29 02:56:02 mail postfix/smtp[1112]: connect to alt1.gmail-smtp-in.l.google.com[2a00:1450:4013:c16::1b]:25: Cannot assign requested address
Sep 29 02:56:32 mail postfix/smtp[1112]: connect to alt1.gmail-smtp-in.l.google.com[142.250.153.27]:25: Connection timed out
Sep 29 02:56:32 mail postfix/smtp[1112]: connect to alt2.gmail-smtp-in.l.google.com[2a00:1450:4025:c03::1b]:25: Cannot assign requested address
Sep 29 02:56:32 mail postfix/smtp[1112]: BFBD0920218: to=<sanjuan.thibaut@gmail.com>, relay=none, delay=60, delays=0.01/0.01/60/0, dsn=4.4.1, status=deferred (connect to alt2.gmail-smtp-in.l.google.com[2a00:1450:4025:c03::1b]:25: Cannot assign requested address)
```

Looks like the issue there is it attempts to connect to Gmail MTA over port 25 and fails. Probably your ISP for that DMS host has blocked outbound traffic on port 25, [this is a common problem we highlight in our debugging docs](https://docker-mailserver.github.io/docker-mailserver/latest/config/debugging/#mail-sent-from-dms-does-not-arrive-at-destination). You'll get that with any mail software.

If you're unable to get that unblocked, you'll want to relay mail through a service like SendGrid.

---

>  I'd like to expose only the Traefik container on the Internet, and route ports through it.

I understand and agree with your reasoning for wanting that. However, it's not clear what the configuration issue is, and I don't have the time to try figure it out for free 😅 

What we have established is this is Traefik specific. DMS is functioning correctly without it. You'll probably run into similar problem with any self-host alternative to DMS too, unless they've done something differently.

If you do have success elsewhere or figure out the problem, do let us know :)

**Comment by sanjuant** [comment]
Ah damn, you're right, my box's router blocks port 25 on exit even though I've opened it in the administration area. "For security they say", long live web neutrality -_-'
![image](https://github.com/docker-mailserver/docker-mailserver/assets/27998877/a627ae4d-e871-4b9c-ab07-4d0cb5eb39d1)

Eventually I'm going to use an ISP that allows server drop-off, so I shouldn't have this problem. 

I'm just annoyed to be able to test before the deposit. Thanks for your help.
